src/mo_frontend/typing.ml

```diff
@@ -1420,15 +1421,15 @@ and infer_exp'' env exp : T.typ =
         scopes = T.ConEnv.add c exp.at env.scopes } in
     let t = infer_exp env' exp1 in
     let t' = T.open_ [t1] (T.close [c] t)  in
-    if not (T.shared t') then
+    if s = Type.Fut && not (T.shared t') then
```

Comment:
Not sure whether `s = Type.Fut &&` is needed if `async*` always has a shared type.

Reply:
I would like to remove the restriction for both types, but should do that in a separate PR. Thanks for finding this.

Reply:
Reverted